lab2/part2/myrouter_part2.py

```diff
@@ -0,0 +1,443 @@
+        self.map   = {}
+    
+    def __iter__(self):
+        return iter(self.map)
+
+    def __getitem__(self, item):
+        if item in self.map and isinstance(self.map[item], ARPContext.ARPEntry):
+            return self.map[item].mac_addr
+        else:
+            return None
+
+    def __setitem__(self, ip_addr, mac_addr):
+        self.add_mapping(ip_addr, mac_addr)
+
+    '''
+    add_mapping
+    Add or updates an IP to MAC address mapping
+      ip_addr      IPv4Address
+      mac_addr     EthAddr
+    '''
+    def add_mapping(self, ip_addr, mac_addr):
+        if isinstance(ip_addr, IPv4Address) and isinstance(mac_addr, EthAddr):
+            self.map[ip_addr] = ARPContext.ARPEntry(mac_addr)
+
+    '''
+    handle_arp_request
+    Process an ARP requests in the provided packet. Return a reply packet
+    if the target IP address is in our table
+       arp_head         Arp packet header to process
+    Returns an ARP reply packet
+    '''
+    def handle_arp_request(self, arp_head: Arp):
+        reply_ip  = arp_head.targetprotoaddr
+        reply_mac = self.map[reply_ip].mac_addr
+
+        #Construct reply packet
+        etp = Ethernet(
+                    ethertype = EtherType.ARP,
+                    src       = reply_mac,
+                    dst       = arp_head.senderhwaddr
+        )
+
+        arp = Arp(
+                operation       = ArpOperation.Reply,
+                senderhwaddr    = reply_mac,
+                senderprotoaddr = reply_ip,
+                targethwaddr    = arp_head.senderhwaddr,
+                targetprotoaddr = arp_head.senderprotoaddr
+        )
+
+        return_pkt = etp + arp
+        return return_pkt
+
+    '''
+    handle_arp_reply
+    Process an ARP reply in the provided packet.
+       arp_head         Arp packet header to process
+    '''
+    def handle_arp_reply(self, arp_head: Arp):
+        self.add_mapping(arp_head.senderprotoaddr, arp_head.senderhwaddr)
+
+    '''
+    get_arp_request
+    Generate an ARP request packet
+      target_addr       Target IP address
+      intf              Outoing interface
+    Returns a new Arp reqeust packet
+    '''
+    def get_arp_request(self, target_addr: IPv4, intf: Interface):
+
+        arp_head = Arp(
+                    operation       = ArpOperation.Request,
+                    senderhwaddr    = intf.ethaddr,
+                    senderprotoaddr = intf.ipaddr,
+                    targetprotoaddr = target_addr
+        )
+        
+        eth_head = Ethernet(
+                        ethertype = EtherType.ARP,
+                        src       = intf.ethaddr,
+                        dst       = SpecialEthAddr.ETHER_BROADCAST.value
+        )
+
+        return_pkt = eth_head + arp_head
+        return return_pkt
+
+    '''
+    Class:          ARPContext.ArpEntry
+    Description:    Table entry pairing a timestamp with a MAC address
+    '''
+    class ARPEntry:
+        def __init__(self, mac_addr, time=time.time()):
```

Comment:
The way this `__init__` is defined implies that instantiations of `ARPEntry` objects without the `time` parameter being supplied will all have the same `time`, which is about the time when the python interpreter initially loaded in this function definition. To fix this, `time=None` can be substituted, and within the function, you can check set `self.time=time.time()` when `time=None`

From https://docs.python.org/3/reference/compound_stmts.html#function-definitions :

>Default parameter values are evaluated from left to right when the function definition is executed. This means that the expression is evaluated once, when the function is defined, and that the same “pre-computed” value is used for each call. This is especially important to understand when a default parameter is a mutable object, such as a list or a dictionary: if the function modifies the object (e.g. by appending an item to a list), the default value is in effect modified. This is generally not what was intended. A way around this is to use None as the default, and explicitly test for it in the body of the function, e.g.: